### PR TITLE
Feat/upload file

### PR DESCRIPTION
**Pull request by Ciszak-Krzysztof**

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -58,6 +58,7 @@ model Recipe {
   authorId         String
   author           User              @relation(fields: [authorId], references: [id])
   ratings          Rating[]
+  imageKey          String            @default("")
```
Comment (edraven-dev) [inline prisma/schema.prisma]: Formatting is shifted a bit here :)

File: src/files/dto/upload-file.dto.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+export class UploadFileDto {
+  @ApiProperty({ type: 'string' })
+  type: string;
+
+  @ApiProperty({ type: 'boolean' })
+  isPublic: boolean;
+}
```
Comment (edraven-dev) [inline src/files/dto/upload-file.dto.ts]: If it is not a response DTO, it should contains some validation eg. `@IsBoolean()`

File: src/files/files.controller.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    description: 'The key of the file to delete',
+    example: '1234567890abcdef',
+  })
+  @ApiOkResponse({
+    description: 'The deleted file',
+    type: DeleteFileResponseDto,
+  })
+  async deleteFile(@Param('key') key: string) {
+    return this.filesService.deleteFile(key);
+  }
```
Comment (edraven-dev) [inline src/files/files.controller.ts]: return types are missing in all the methods

File: src/files/files.service.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  UploadFileResponseDto,
+} from './dto';
+import { UploadFileDto } from './dto/upload-file.dto';
+
+@Injectable()
+export class FilesService {
+  private client: S3Client;
+  private bucketName = this.configService.get('AWS_S3_BUCKET');
+
+  constructor(private configService: ConfigService) {
```
Comment (edraven-dev) [inline src/files/files.service.ts]:
```suggestion
  constructor(private readonly configService: ConfigService) {
```

File: src/files/files.service.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    this.client = new S3Client({
+      region: s3_region,
+      credentials: {
+        accessKeyId: this.configService.get('AWS_ACCESS_KEY'),
+        secretAccessKey: this.configService.get('AWS_SECRET_ACCESS_KEY'),
+      },
+      forcePathStyle: true,
+    });
```
Comment (edraven-dev) [inline src/files/files.service.ts]: you would get troubles with testing class with the dependency instantiated in the constructor. in NestJS you can create own providers which are instantiated by the framework and stored in IoC. So, in your module, you can add to the providers array a new object with some injection token and `useFactory` method which should instantiate the S3Client. then you can inject that provider in this class using `@Inject(<injection_token_of_s3_client_provider>)` decorator

File: src/recipes/recipes.service.ts [hunk 6]
```diff
@@ -11,10 +11,14 @@ import {
...
 } from './dto/get-recipe-response.dto';
 import { AddRecipeDto } from './dto/add-recipe.dto';
+import { FilesService } from '@/files/files.service';
 
 @Injectable()
 export class RecipesService {
-  constructor(private recipeRepository: RecipesRepository) {}
+  constructor(
+    private recipeRepository: RecipesRepository,
+    private filesService: FilesService,
```
Comment (edraven-dev) [inline src/recipes/recipes.service.ts]:
```suggestion
    private readonly recipeRepository: RecipesRepository,
    private readonly filesService: FilesService,
```

File: src/recipes/recipes.service.ts [hunk 7]
```diff
@@ -40,9 +44,20 @@ export class RecipesService {
...
     authorId: string,
     addRecipeDto: AddRecipeDto,
+    file?: Express.Multer.File,
   ): Promise<GetRecipeResponse> {
+    const fileUploadResult =
+      file &&
+      (await this.filesService.uploadSingleFile({
+        file,
+        uploadFileDto: { type: 'recipe', isPublic: false },
+      }));
```
Comment (edraven-dev) [inline src/recipes/recipes.service.ts:55]: it would be clearer what it does if it would be extracted to separated method ;)

**Review by edraven-dev — APPROVED** [review]

File: src/config/swagger.config.ts [hunk 8]
```diff
@@ -3,8 +3,8 @@ import { INestApplication } from '@nestjs/common';
 
 export function setupSwagger(app: INestApplication): void {
   const config = new DocumentBuilder()
-    .setTitle('Users')
-    .setDescription('The Users API description')
+    .setTitle('Leftovers API')
+    .setDescription('The Leftovers API description')
     .setVersion('0.1')
```
Comment (rad20c) [inline src/config/swagger.config.ts]: You can sync displayed version here with the version in `package.json`:

1. Add `"resolveJsonModule": true` in `tsconfig.json` under `compilerOptions`
2. Add `import { version } from '../../package.json';`
3. Replace `.setVersion('0.1')` with `.setVersion(version)`

File: src/files/dto/index.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export * from './get-file-url-response.dto';
+export * from './upload-file-response.dto';
+export * from './delete-file-response.dto';
```
Comment (rad20c) [inline src/files/dto/index.ts:3]: In `.vscode/settings.json` you can add this for automatic import organization on save:

```
  "editor.codeActionsOnSave": {
    "source.organizeImports": "explicit"
  },
```

File: src/files/files.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+      ],
+      providers: [FilesService],
+    }).compile();
+
+    service = module.get<FilesService>(FilesService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
```
Comment (rad20c) [inline src/files/files.service.spec.ts:30]: tests coming in next PR?

File: src/files/files.service.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+} from './dto';
+import { UploadFileDto } from './dto/upload-file.dto';
+
+@Injectable()
+export class FilesService {
+  private client: S3Client;
+  private bucketName = this.configService.get('AWS_S3_BUCKET');
+
+  constructor(private configService: ConfigService) {
+    const s3_region = this.configService.get('AWS_BUCKET_REGION');
```
Comment (rad20c) [inline src/files/files.service.ts]: should we stick to camelCase?
```suggestion
    const s3Region = this.configService.get('AWS_BUCKET_REGION');
```

**Review by rad20c — APPROVED** [review]

File: src/files/files.service.spec.ts (hunk 10 above)
Comment (Ciszak-Krzysztof) [reply src/files/files.service.spec.ts:30]: yes

**Pull request merged**